File: project_euler/problem_447/solution.py
"""
Problem 447: Retractions B

The number of retractions R(n) for affine maps f(x) = ax + b mod n equals
the number of pairs (a, b) with a^2 ≡ a (mod n) and ab ≡ 0 (mod n).

Key result: R(n) = prod_{p^e || n} (1 + p^e), a multiplicative function.

We compute sum_{n=2}^{N} R(n) mod (10^9 + 7) using a linear sieve.
"""

from math import gcd

MOD = 10**9 + 7

# --- Method 1: Linear sieve using smallest prime factor ---
def solve_linear_sieve(N: int, mod: int):
    """Compute R(n) for n=1..N via linear sieve. Return (R_array, total_sum)."""
    spf = list(range(N + 1))  # smallest prime factor
    for i in range(2, int(N**0.5) + 1):
        if spf[i] == i:  # i is prime
            for j in range(i * i, N + 1, i):
                if spf[j] == j:
                    spf[j] = i

    R = [0] * (N + 1)
    R[1] = 1  # R(1) = 1 (only a=0, b=0 trivially, but by convention)
    total = 0

    for n in range(2, N + 1):
        p = spf[n]
        # Find p^e exactly dividing n
        e = 0
        m = n
        while m % p == 0:
            m //= p
            e += 1
        # R(n) = R(m) * (1 + p^e)
        pe = p ** e
        R[n] = R[m] * ((1 + pe) % mod) % mod
        total = (total + R[n]) % mod

    return R, total

# --- Method 2: Brute force for small n (verification) ---
def R_brute(n: int):
    """Count retractions by brute force: pairs (a, b) with a^2≡a, ab≡0 mod n."""
    count = 0
    for a in range(n):
        if (a * a) % n != a % n:
            continue
        for b in range(n):
            if (a * b) % n == 0:
                count += 1
    return count

# --- Method 3: Product formula (verification) ---
def R_product(n: int):
    """Compute R(n) = prod_{p^e || n} (1 + p^e)."""
    if n <= 1:
        return 1
    result = 1
    d = 2
    temp = n
    while d * d <= temp:
        if temp % d == 0:
            pe = 1
            while temp % d == 0:
                pe *= d
                temp //= d
            result *= (1 + pe)
        d += 1
    if temp > 1:
        result *= (1 + temp)
    return result

# --- Compute answer ---
N = 10**7
R_arr, ans = solve_linear_sieve(N, MOD)

# --- Verify against brute force for small n ---
for n in range(2, 50):
    bf = R_brute(n)
    pf = R_product(n)
    assert bf == pf, f"Brute vs product mismatch at n={n}: {bf} vs {pf}"
    assert bf % MOD == R_arr[n], f"Sieve mismatch at n={n}: {R_arr[n]} vs {bf}"

# --- Verify product formula examples ---
assert R_product(2) == 3
assert R_product(6) == 12
assert R_product(12) == 20
assert R_product(30) == 72

print(ans)

File: project_euler/problem_447/solution.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Problem 447: Retractions B
 *
 * R(n) = number of affine retractions f(x) = ax + b (mod n), i.e.,
 *        pairs (a,b) with a^2 ≡ a (mod n) and ab ≡ 0 (mod n).
 *
 * Key formula: R(n) = prod_{p^e || n} (1 + p^e).
 * R is multiplicative, computed via linear sieve.
 *
 * Answer: sum_{n=2}^{10^7} R(n) mod (10^9 + 7).
 */

const long long MOD = 1e9 + 7;
const int N = 10000000;

int spf[N + 1];         // smallest prime factor
long long R_val[N + 1]; // R(n) mod MOD

void compute_spf() {
    for (int i = 0; i <= N; i++) spf[i] = i;
    for (int i = 2; (long long)i * i <= N; i++) {
        if (spf[i] == i) { // i is prime
            for (int j = i * i; j <= N; j += i) {
                if (spf[j] == j) spf[j] = i;
            }
        }
    }
}

int main() {
    compute_spf();

    R_val[1] = 1;
    long long total = 0;

    for (int n = 2; n <= N; n++) {
        int p = spf[n];
        // Extract p^e from n
        int m = n;
        long long pe = 1;
        while (m % p == 0) {
            m /= p;
            pe *= p;
        }
        // R(n) = R(m) * (1 + p^e) mod MOD
        R_val[n] = R_val[m] % MOD * ((1 + pe) % MOD) % MOD;
        total = (total + R_val[n]) % MOD;
    }

    cout << total << endl;

    // Verification for small cases
    // R(2) = 3, R(3) = 4, R(4) = 5, R(6) = 12
    assert(R_val[2] == 3);
    assert(R_val[3] == 4);
    assert(R_val[4] == 5);
    assert(R_val[6] == 12);
    assert(R_val[12] == 20);
    assert(R_val[30] == 72);

    return 0;
}
